### PR TITLE
feat(Core/Hooks): added hooks from MS

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## CHANGES PROPOSED:
- Add hooks from MaelstromCore
- General changes for support module `mod-solo-3v3-arena`
- After this commit @Yehonal maybe give you all a module `mod-solo-3v3-arena` :D

## TESTS PERFORMED:
- Build only

## HOW TO TEST THE CHANGES:
- Test bg/arena (enter/progress and etc)
- Test all charters (guild, arena)
- Test spells (basic test, casts, buffs, etc)
- Test phase (gm mode and other)

## KNOWN ISSUES AND TODO LIST:
- 

## Target branch(es):
- [x] Master

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by BarbzYHOOL** [comment]
put a title xD

**Comment by Winfidonarleyan** [comment]
Yes yes

**Comment by merdlin** [comment]
what is this PR for?

**Comment by Winfidonarleyan** [comment]
For module 3v3 solo

File: src/server/game/Petitions/PetitionMgr.h
Comment (IntelligentQuantum) [inline src/server/game/Petitions/PetitionMgr.h:28]: Do we need this?

Comment (Winfidonarleyan) [reply src/server/game/Petitions/PetitionMgr.h:28]: Now no

**Comment by merdlin** [comment]
Tested, not found errors.

**Comment by Winfidonarleyan** [comment]
Add information for PR

**Comment by merdlin** [comment]
I tested again, I found no error. successful build.

**Comment by Kitzunu** [comment]
Please fix Codacy https://app.codacy.com/manual/FrancescoBorzi/azerothcore-wotlk/pullRequest?prid=5541563

File: src/server/game/Battlegrounds/BattlegroundMgr.cpp
Comment (FrancescoBorzi) [inline src/server/game/Battlegrounds/BattlegroundMgr.cpp:773]: why? we should do the opposite actually...

File: src/server/game/Entities/Item/Item.cpp
Comment (FrancescoBorzi) [inline src/server/game/Entities/Item/Item.cpp:94]: it's not clear what this `if` is controlling. So please use brackets and indentation

File: src/server/game/Entities/Player/Player.cpp
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp:3361]: why not add a condition to the existing `if` instead of adding a new `if` block?

File: src/server/game/Entities/Player/Player.h
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.h:1945]: why removing `[[nodiscard]]` ???

File: src/server/game/Entities/Unit/Unit.cpp
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:19804]: same as above: use brackets **and** indentation, otherwise it's confusing

File: src/server/game/Guilds/Guild.h
Comment (FrancescoBorzi) [inline src/server/game/Guilds/Guild.h:758]: missing `[[nodiscard]]`

File: src/server/game/Handlers/PetitionsHandler.cpp
Comment (FrancescoBorzi) [inline src/server/game/Handlers/PetitionsHandler.cpp:880]: can use `auto` in cases like this

File: src/server/game/Handlers/PetitionsHandler.cpp
Comment (FrancescoBorzi) [inline src/server/game/Handlers/PetitionsHandler.cpp:888]: what's the meaning of this change?

File: src/server/game/Battlegrounds/ArenaTeam.cpp
Comment (FrancescoBorzi) [inline src/server/game/Battlegrounds/ArenaTeam.cpp:981]: can use `auto` in cases like this, no reason to repeat `uint32` twice in the same line

File: src/server/game/Scripting/ScriptMgr.cpp
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.cpp:2201]: can be just `return false` here

Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.cpp:2201]: are we returnig `false` if the LAST value of `FOR_SCRIPTS_RET` is false or if ANY of them returns false?

File: src/server/game/Scripting/ScriptMgr.cpp
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.cpp:2218]: are we returnig `false` if the LAST value of `FOR_SCRIPTS_RET` is false or if ANY of them returns false?

File: src/server/game/Scripting/ScriptMgr.cpp
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.cpp:2229]: are we returnig `false` if the LAST value of `FOR_SCRIPTS_RET` is false or if ANY of them returns false?

File: src/server/game/Scripting/ScriptMgr.cpp
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.cpp:3283]: this file is getting HUGE, wouldn't be better to split it?

File: src/server/game/Scripting/ScriptMgr.h
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.h:1038]: fix indentation and add `[[nodiscard]]` to all these methods please

File: src/server/game/Scripting/ScriptMgr.h
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.h:1205]: `[[nodiscard]]`

File: src/server/game/Scripting/ScriptMgr.h
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.h:1310]: `[[nodiscard]]` to all getter methods please

File: src/server/game/Scripting/ScriptMgr.h
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.h:1357]: `[[nodiscard]]` to all getter methods please

File: src/server/scripts/Outland/BlackTemple/boss_mother_shahraz.cpp
Comment (FrancescoBorzi) [inline src/server/scripts/Outland/BlackTemple/boss_mother_shahraz.cpp:269]: brackets

File: src/cmake/ac_macros.cmake
Comment (Yehonal) [inline src/cmake/ac_macros.cmake:9]: Just as a reminder for other devs: this can be used by modules to automatically load game scripts from a gived folder inside the **scripts_STAT_SRCS** array (without manually add them one by one).

Winfy could you give me edit permissions here, thus I can add the description of what it does inside the cmake comments?

File: src/server/game/Scripting/ScriptMgr.cpp
Comment (Yehonal) [reply src/server/game/Scripting/ScriptMgr.cpp:3283]: Better to open another PR/issue for it. We cannot increase the complexity of this one now

File: src/cmake/ac_macros.cmake
Comment (Winfidonarleyan) [reply src/cmake/ac_macros.cmake:9]: yes, sec

File: src/server/game/Scripting/ScriptMgr.h
Comment (Winfidonarleyan) [reply src/server/game/Scripting/ScriptMgr.h:1357]: I may have misunderstood, but I'd like to see an example.

**Comment by Winfidonarleyan** [comment]
Maybe need retest?

File: src/server/game/Scripting/ScriptMgr.h
Comment (FrancescoBorzi) [reply src/server/game/Scripting/ScriptMgr.h:1357]: image

**Review by Yehonal — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers can anyone please test this PR?

**Comment by pangolp** [comment]
Hello, good afternoon. I share the captures of the tests you make, if you think I lack something to test, let me know and I will make the tests.

WoWScrnShot_123020_170642

WoWScrnShot_123020_170730

WoWScrnShot_123020_170800

WoWScrnShot_123020_171657

WoWScrnShot_123020_173903

**Review by pangolp — APPROVED** [review]

**Comment by peti446** [comment]
Hi, I am unable to test this until Sunday but I would like to test it with 1v1 module as it does some changes to the script calls before we merge it

**Comment by Yehonal** [comment]
Yes please guys, test the Arenas they are very important.

This patch is a job I requested that will allow me to share part of the AzerothShard custom modules, including 1v1 and 3v3soloq (reusing the 4v4 unused bracket instead of the 5v5)...but also the time walking and tons of other cool features.

So let's make sure that it doesn't break the core features as soon as possible, otherwise, we will merge it anyway to avoid new merge conflicts in future

**Comment by FrancescoBorzi** [comment]
@peti446 sure, we will wait then ;)

**Comment by peti446** [comment]
@FrancescoBorzi This will trigger assertions with the 1v1 and probally crash or cause all sort of glishes during gameplay in release, for now I identified one problem: SlotByType it will return 255 instead of the slot specified in the 
 1v1 arena (1 in this case) the core will then access unspecified memory returning garbage. With a quick some or most of the changes made by the pr https://github.com/azerothcore/azerothcore-wotlk/pull/3508 have been reverted. 

However, if no module that uses the new hooks provided in the above PR is used this works fine.

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @pklloveyou can anyone please have a final check on this before we merge it? It was tested in past, but then it was updated after the test

**Comment by FrancescoBorzi** [comment]
waiting for a review from @Yehonal

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan the build is now broken :(

**Comment by Winfidonarleyan** [comment]
Okay, need time to fix

**Pull request merged**
